src/parquet_writer.h

```diff
@@ -0,0 +1,190 @@
+        os << row.Get(i.key).ToNumber().DoubleValue();
+        break;
+
+      case parquet::Type::BOOLEAN:
+        os << row.Get(i.key).ToBoolean().Value();
+        break;
+
+      case parquet::Type::BYTE_ARRAY:
+        os << row.Get(i.key).ToString().Utf8Value();
+        break;
+
+      default:
+        os << "null";
+      }
+    }
+    os << parquet::EndRow;
+
+    return Napi::Boolean::New(env, true);
+  }
+
+  Napi::Value open(const Napi::CallbackInfo& info) {
+    // Open file
+    std::shared_ptr<arrow::io::FileOutputStream> out_file;
+    PARQUET_ASSIGN_OR_THROW(
+      out_file,
+      arrow::io::FileOutputStream::Open(filepath)
+    );
+
+    // Setup output stream
+    os = parquet::StreamWriter {parquet::ParquetFileWriter::Open(
+      out_file,
+      schema,
+      builder.build()
+    )};
+
+    return Napi::Boolean::New(info.Env(), true);
+  }
+
+  Napi::Value setRowGroupSize(const Napi::CallbackInfo& info) {
+    builder.max_row_group_length(info[0].ToNumber().Int64Value());
+    return Napi::Boolean::New(info.Env(), true);
+  }
+
+protected:
+
+  std::string filepath;
+  std::shared_ptr<parquet::schema::GroupNode> schema;
+  parquet::WriterProperties::Builder builder;
+  parquet::StreamWriter os;
+  std::vector<Column> columns;
+
+  parquet::ConvertedType::type convertedTypeFromString(const std::string& type) {
+    // Maybe it would be faster if it hashed the string and used a switch statement
+    if (type == "INT32") {
+      return parquet::ConvertedType::INT_32;
+    } else if (type == "INT64") {
+      return parquet::ConvertedType::INT_64;
+    } else if (type == "TIMESTAMP_MICROS") {
+      return parquet::ConvertedType::INT_64;
+    } else if (type == "TIMESTAMP_MILLIS") {
+      return parquet::ConvertedType::INT_64;
+    } else if (type == "DATE32") {
+      return parquet::ConvertedType::INT_32;
+    } else if (type == "DOUBLE") {
+      return parquet::ConvertedType::NONE; // Looks weird but is correct
+    } else if (type == "BOOLEAN") {
+      return parquet::ConvertedType::NONE; // ditto
+    } else if (type == "UTF8") {
+      return parquet::ConvertedType::UTF8;
+    }
+
+    throw std::invalid_argument("Unsupported parquet field type");
+  }
+
+  parquet::Type::type logicalTypeFromString(const std::string& type) {
+    if (type == "INT32") {
+      return parquet::Type::INT32;
+    } else if (type == "INT64") {
+      return parquet::Type::INT64;
+    } else if (type == "TIMESTAMP_MICROS") {
+      return parquet::Type::INT64;
+    } else if (type == "TIMESTAMP_MILLIS") {
+      return parquet::Type::INT64;
+    } else if (type == "DATE32") {
+      return parquet::Type::INT32;
+    } else if (type == "DOUBLE") {
+      return parquet::Type::DOUBLE;
+    } else if (type == "BOOLEAN") {
+      return parquet::Type::BOOLEAN;
+    } else if (type == "UTF8") {
+      return parquet::Type::BYTE_ARRAY;
+    }
+
+    throw std::invalid_argument("Unsupported parquet field type");
+  }
```

Comment:
These 2 functions are pure, they shouldn't be in a class.

Reply:
Also if you throw C++ errors, you need to catch them before they reach JS, otherwise JS won't be able to recover from these with `try/catch`.

Reply:
Making the functions static. Organizationally they should be part of the class, but they definitely don't need to be member functions. Static splits the difference.

I'll catch the c++ exceptions and throw js exceptions as a result.